pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,178 @@
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a new cache to watch on Route objects from every namespace.
+	newCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cache to the manager so that the cache is started along with the other runnables.
+	mgr.Add(newCache)
+	reconciler := &reconciler{
+		cache:     newCache,
+		Namespace: namespace,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCache),
+		handler.EnqueueRequestsFromMapFunc(reconciler.routeToIngressController)); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeToIngressController creates a reconcile.Request for all the Ingress Controllers related to the Route object.
+func (r *reconciler) routeToIngressController(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ic := range routeObject.Status.Ingress {
+		log.Info("queueing ingresscontroller", "name", ic.RouterName)
+		// Create a reconcile.Request for the Ingress Controller.
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      ic.RouterName,
+				Namespace: r.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
```

Comment:
Is the shard name in the `request.Name`?

Reply:
The `types.Namespace` struct is embedded inside the `reconcile.Request` struct. The fields and the methods of the embedded struct can be directly accessed without using the embedded struct name. The details of embedding in golang can be found here: https://go.dev/doc/effective_go#embedding

Reply:
I don't think we should be deleting the metrics, but if we do, how about using the ingress controller name?

Reply:
The reconcile.Request is for the Ingress Controller only. So we're using the Ingress Controller Name.

The reason we need to delete metrics is that if not done so, it'll keep on exporting stale metrics. We need to delete the metrics of an Ingress Controller whenever it gets deleted.

Reply:
Prometheus will continue to export these metrics for an ingress controller after the router gets deleted?  I don't see how that could happen, unless this code introduces that buggy behavior.  If there's no instance to scrape, how would it get repeatedly written?

Reply:
The scraping happens for the metrics being exported here by the Cluster Ingress Operator and not the Router. If the specific Shard's metric information is not deleted, the Cluster Ingress Operator will still continue to export the last stored value for that Shard.

Reply:
Here's an example of metrics being deleted when Ingress Controller is deleted: https://github.com/openshift/cluster-ingress-operator/blob/79778714fb968876eff923a870dd8fe56ed723cb/pkg/operator/controller/ingress/controller.go#L849-L850
If not deleted stale metrics would still get exported.

Reply:
I still don't feel it's helpful to delete these metrics, but at least it is more correct to do it in the deletion finalizer.

Reply:
For example, if there are 3 shards `shard0`, `shard1` and `shard2` having 10, 5 and 3 routes assigned to them respectively. The metrics that will be exported will have label `shard0`, `shard1` and `shard2` and the corresponding values will be 10, 5 and 3. Now, let's say if `shard0` is deleted and if we don't delete the corresponding metric label then the metrics that will be exported will have label `shard0`, `shard1` and `shard2` and the corresponding values will be 10, 5 and 0. This is incorrect as `shard2` is non-existent and it will impact the overall indicators like `min`, `avg` as well as indicators for the range of 0 to 33% of total routes. Thus, we need to delete the corresponding label of `shard2`.

Reply:
> If the specific Shard's metric information is not deleted, the Cluster Ingress Operator will still continue to export the last stored value for that Shard.

This is a bug.  It should not export the last stored value for a metric that has not been refreshed.

Reply:
We are updating the metric values based on events. That's why we need to delete the last stored value corresponding to a deleted Ingress Controller from the metric when it is deleted. That's how we have to refresh the data. If we don't then it will give rise to a bug.

> Now, let's say if `shard0` is deleted and if we don't delete the corresponding metric label then the metrics that will be exported will have label `shard0`, `shard1` and `shard2` and the corresponding values will be `10`, `5` and `0`.

As we are deleting the corresponding metric label then the metrics that will be exported will have label `shard0` and `shard1` and the corresponding values will be `10` and `5`. The metrics will not have the label `shard2` as it is deleted. I hope this clarifies confusion.

Reply:
Rather, it should not export the last stored value for an ingress controller that is deleted.  It should stop exporting, and let the time series age out over time.

Reply:
We are using `GaugeVec` which stores the metric values and every time the metric endpoint is scraped for metrics, all the stored values are exported. As this is event based, we are storing and removing metric values from the `GaugeVec`. If we don't explicitly remove any label it'll export all data points. There's no way of excluding any stored value from being exported. That's why it has to be removed explicitly when the IC is removed. I am also following the similar pattern used in other parts of Cluster Ingress Operator as shown here:
https://github.com/openshift/cluster-ingress-operator/blob/79778714fb968876eff923a870dd8fe56ed723cb/pkg/operator/controller/ingress/controller.go#L849-L850

If we want to just export the current information then we have to gather and export from scratch all the metrics information every time the metrics endpoint is scraped. We cannot use `GaugeVec` and have to write a separate exporter module. This will always export the current information as it will fetch every time. Thus, automatically excluding any IC which is deleted.  I had proposed this as Design 1 in the design doc: https://docs.google.com/document/d/1Ojlpm2OTmGPY-5aHcP80HF7deBf1DRWARLrsSkrWneA/edit#heading=h.20vwkwvvqct1

As we are going ahead with Design 2, which gathers metrics based on events, the data has to be explicitly removed.

Reply:
Please move it from here to the deletion finalizer.  We may need to look into why this pattern is perpetuating.  The first time it was used, maybe it shoudn't have been.  We don't do deletion for all metrics and we haven't had a problem with that in other cases.

Reply:
Moved the deletion of the metric label to finalizer.

Reply:
I have asked about this in forum-monitoring: https://coreos.slack.com/archives/C0VMT03S5/p1663614322572739

Reply:
@arkadeepsen, Grant added some logic to remove a router from routes' statuses when the ingresscontroller for the router is deleted; see <https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/router_status.go>.  In your example, if the cluster admin deletes shard0, the operator's ingress controller should remove shard0 from the status of all routes that were in shard0.  Then I'd expect the route-metrics controller to observe those route status updates and decrement shard0's gauge to 0.  Is that not happening, or is it insufficient?

Reply:
> In your example, if the cluster admin deletes shard0, the operator's ingress controller should remove shard0 from the status of all routes that were in shard0. Then I'd expect the route-metrics controller to observe those route status updates and decrement shard0's gauge to 0. Is that not happening, or is it insufficient?

@Miciah, If the call to the `DeleteRouteMetricsControllerRoutesPerShardMetric` is removed from the `Reconcile` function, then `shard0`'s gauge will be decremented to 0. But only decrementing the gauge would not be correct. When `shard0` is deleted, then we should not be exporting the metrics for this shard at all. If we only let it get decremented to 0 and not delete it, the issue is that it'll be exporting stale metrics of `shard0` which is non-existent. This will impact the indicators like `min`, `avg` and `median` of `RoutesPerShard` metric.

Reply:
> I have asked about this in forum-monitoring: https://coreos.slack.com/archives/C0VMT03S5/p1663614322572739

@candita I think there's some difference in understanding what is being deleted. In the question you have mentioned whether a metric should be `specifically deleted from Prometheus when its OpenShift object is deleted`. The deletion is not happening from the Prometheus backend, rather it is being deleted from the `GaugeVec` variable local to the Cluster Ingress Operator. Every time the metrics endpoint of Cluster Ingress Operator is scraped for fetching metrics, the data stored in the local `GaugeVec` variable is exported. Here is where the problem lies. If the stale information is not removed from this local `GaugeVec` variable it will continue to get exported by the Cluster Ingress Operator's metrics endpoint. That's why whenever the Ingress Controller is deleted, the corresponding label for that Ingress Controller should also be deleted from the local `GaugeVec` variable. Once this label is deleted then the corresponding data for the Ingress Controller will not be exported. Prometheus will not write back any old data and it will only reflect the data that is being exported by the Cluster Ingress Operator, whether it's correct or stale data.

Reply:
I'll defer to the experts here, and the response from the monitoring team on that Slack thread says that if we're not using the custom collector pattern, then "the exporter should delete metrics associated to an entity (label set) when it knows that the entity is gone".  That response and Arkadeep's explanation make sense to me:  The operator should stop exporting metrics for a shard once the shard is gone.